pkg/kv/kvserver/replica_application_result.go

```diff
@@ -505,6 +507,77 @@ func (r *Replica) stagePendingTruncationRaftMuLocked(pt pendingTruncation) {
 	r.asLogStorage().stagePendingTruncationRaftMuLocked(pt)
 }
 
+func (r *replicaLogStorage) stagePendingTruncationSharedRaftMuLockedMuLocked(pt pendingTruncation) {
+	// This code is shared between regular log truncations and snapshot-induced
+	// truncations.
+	r.shMu.trunc = pt.RaftTruncatedState
+	// Ensure the raft log size is not negative since it isn't persisted between
+	// server restarts.
+	// TODO(pav-kv): should we distrust the log size if it goes negative?
+	r.shMu.size = max(r.shMu.size+pt.logDeltaBytes, 0)
+	r.shMu.lastCheckSize = max(r.shMu.lastCheckSize+pt.logDeltaBytes, 0)
+	if !pt.isDeltaTrusted {
+		r.shMu.sizeTrusted = false
+	}
+}
+
+func (r *replicaLogStorage) stagePendingTruncationOnSnapshotRaftMuLocked(
+	truncState kvserverpb.RaftTruncatedState,
+) {
+	r.raftMu.AssertHeld()
+
+	// A snapshot application implies a log truncation to the snapshot's index,
+	// and we apply the resulting memory state here (before the snapshot takes
+	// effect, i.e. the log entries disappear from storage). This avoids
+	// situations in which entries were already removed, but the in-mem state
+	// indicates that they ought to still exist.
+	//
+	// The truncation finalized below, after the snapshot is visible.
+
+	// TODO(during review): when is it actually safe to clear the cache if we
+	// don't want to do it under both mutexes? The main addition to the cache does
+	// hold raftMu (append) but there are various paths into the cache coming
+	// through LoadEntries and via term loading. It would be easier if we only
+	// added to the cache in the one place that seems to matter - entry append.
+	// Is that a change we should make now, while we're here?
+	r.cache.Clear(r.ls.RangeID, math.MaxUint64)
```

Comment:
Were we not clearing the cache before? Why was it ok?

Reply:
we were, here: https://github.com/cockroachdb/cockroach/pull/145947/files#r2088757820

Reply:
Ah, there was a `Drop()` call. Should we call `Drop()` here instead of `Clear()`? The `math.MaxUint64` smells with an overflow danger, since these low/high index methods tend to have +-1 arithmetics.

Reply:
good point, done with next push.